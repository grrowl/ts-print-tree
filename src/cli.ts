#!/usr/bin/env node

import { readFileSync } from "fs";
import { join } from "path";
import { tree, VisibilityLevel, TreeNode } from ".";

const args = process.argv.slice(2);

const defaultIgnore = [
  "node_modules",
  /\.git/,
  /\.vscode/,
  /\.DS_Store/,
  /\.test\.ts$/,
  /\.spec\.ts$/,
];

function printHelp() {
  console.log("Usage: ts-print-tree [options]");
  console.log("--help       Show help");
  console.log("--version    Show version number");
  console.log("--ignore     Ignore files matching the specified patterns");
  console.log("--no-default Don't include default ignore patterns");
  console.log("--cwd        Set the current working directory");
  console.log("--protected  Include protected members");
  console.log("--private    Include private members");
  console.log("--tree       Output as tree (default)");
  console.log("--list       Output as markdown list");
}

function printVersion() {
  const packageJson = JSON.parse(
    readFileSync(join(__dirname, "../package.json"), "utf-8"),
  );
  console.log(`Version: ${packageJson.version}`);
}

export const pathFilter =
  (ignorePatterns: (string | RegExp)[]) => (path: string) =>
    !ignorePatterns.some((pattern) =>
      typeof pattern === "string" ? path.includes(pattern) : pattern.test(path),
    );

function formatNodeName(node: TreeNode): string {
  const visibilityPrefix =
    node.visibility && node.visibility !== "public"
      ? `${node.visibility} `
      : "";
  const typePrefix =
    node.type !== "directory" && node.type !== "file" ? `${node.type} ` : "";
  return `${visibilityPrefix}${typePrefix}${node.name}`;
}

export function formatAsTree(
  node: TreeNode,
  prefix = "",
  isLast = true,
  isRoot = true,
): string {
  const formattedName = formatNodeName(node);
  let output = isRoot
    ? `${formattedName}${node.signature ? node.signature : ""}\n`
    : `${prefix}${isLast ? "└── " : "├── "}${formattedName}${node.signature ? node.signature : ""}\n`;

  if (node.children) {
    node.children.forEach((child, index) => {
      const isLastChild = index === node.children!.length - 1;
      const newPrefix = isRoot ? "" : prefix + (isLast ? "    " : "│   ");
      output += formatAsTree(child, newPrefix, isLastChild, false);
    });
  }
  return output;
}

export function formatAsList(node: TreeNode, depth = 0): string {
  const formattedName = formatNodeName(node);
  let output = `${"  ".repeat(depth)}${depth === 0 ? "" : "- "}${formattedName}${node.signature ? node.signature : ""}\n`;
  if (node.children) {
    node.children.forEach((child) => {
      output += formatAsList(child, depth + 1);
    });
  }
  return output;
}

if (args.includes("--help")) {
  printHelp();
} else if (args.includes("--version")) {
  printVersion();
} else {
  const ignoreIndex = args.indexOf("--ignore");
  let ignorePatterns: (string | RegExp)[] = [];
  let cwd = process.cwd();
  let visibilityLevel = VisibilityLevel.Public;
  let outputFormat: "tree" | "list" = "tree";

  if (ignoreIndex !== -1) {
    ignorePatterns = args
      .slice(ignoreIndex + 1)
      .filter((pat) => !pat.startsWith("--"))
      .map((pat) =>
        pat.startsWith("/") && pat.endsWith("/") && pat.includes("i")
          ? new RegExp(pat.slice(1, -2), "i")
          : pat.startsWith("/") && pat.endsWith("/")
            ? new RegExp(pat.slice(1, -1))
            : pat,
      );
  }

  if (!args.includes("--no-default")) {
    ignorePatterns.push(...defaultIgnore);
  }

  const cwdIndex = args.indexOf("--cwd");
  if (cwdIndex !== -1 && args[cwdIndex + 1]) {
    cwd = args[cwdIndex + 1];
  }

  if (args.includes("--private")) {
    visibilityLevel = VisibilityLevel.Private;
  } else if (args.includes("--protected")) {
    visibilityLevel = VisibilityLevel.Protected;
  }

  if (args.includes("--list")) {
    outputFormat = "list";
  }

  const projectStructure = tree(
    cwd,
    pathFilter(ignorePatterns),
    visibilityLevel,
  );

  const output =
    outputFormat === "tree"
      ? formatAsTree(projectStructure)
      : formatAsList(projectStructure);

  console.log(output);
}
