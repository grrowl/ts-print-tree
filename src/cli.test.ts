import {
  describe,
  expect,
  test,
  jest,
  beforeAll,
  afterAll,
  beforeEach,
} from "@jest/globals";

describe("cli module", () => {
  let mockLog: jest.SpiedFunction<typeof console.log>;
  let originalArgv: string[];

  beforeAll(() => {
    mockLog = jest.spyOn(console, "log").mockImplementation(() => {});
    originalArgv = process.argv;
  });

  afterAll(() => {
    mockLog.mockRestore();
    process.argv = originalArgv;
  });

  beforeEach(() => {
    jest.resetModules();
  });

  test("call with no arguments", () => {
    process.argv = ["node", "script.js"];

    require("./cli");

    expect(console.log).toHaveBeenCalled();
    expect(mockLog.mock.calls).toMatchSnapshot();
  });

  test("call with arguments: ['--list', '--private']", () => {
    process.argv = ["node", "script.js", "--list", "--private"];

    require("./cli");

    expect(console.log).toHaveBeenCalled();
    expect(mockLog.mock.calls).toMatchSnapshot();
  });
});
