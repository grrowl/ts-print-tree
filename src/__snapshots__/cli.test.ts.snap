// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`cli module call with arguments: ['--list', '--private'] 1`] = `
[
  [
    "ts-print-tree/
  - src/
    - cli.ts
      - private const args: string[]
      - private const defaultIgnore: (string | RegExp)[]
      - private function printHelp(): void
      - private function printVersion(): void
      - const pathFilter: (ignorePatterns: (string | RegExp)[]) => (path: string) => boolean
      - private function formatNodeName(node: TreeNode): string
      - function formatAsTree(node: TreeNode, prefix?: string, isLast?: boolean, isRoot?: boolean): string
      - function formatAsList(node: TreeNode, depth?: number): string
    - index.ts
      - private const TS_FILE_MATCH: RegExp
      - interface TreeNode
      - private function readTsConfig(rootDir: string): ParsedCommandLine
      - private function createProgram(rootDir: string): Program
      - private function analyzeFile(sourceFile: SourceFile, typeChecker: TypeChecker, visibilityLevel: VisibilityLevel): TreeNode[]
      - private function traverseDirectory(dir: string, program: Program, pathFilter: (path: string) => boolean, visibilityLevel: VisibilityLevel): TreeNode
      - function tree(rootDir?: string, pathFilter?: (path: string) => boolean, visibilityLevel?: VisibilityLevel): TreeNode
    - tests/
      - class.ts
        - private const tree: "ðŸŒ³"
        - class DefaultClass
          - property static foo: string
          - property bar: string[]
          - method constructor(baz: number[]): DefaultClass
        - class NamedClass extends DefaultClass
          - private property static version: number
          - method get version(): number
          - method constructor(): NamedClass
          - method double(): void
          - private method triple(): void
          - protected method quadruple(): void
          - method octuple(): void
        - const DefaultClass: typeof DefaultClass
      - consts.ts
        - private const privateString: string
        - private const privateNumber: number
        - private const privateBoolean: boolean
        - const publicArray: string[]
        - interface PublicInterface
        - private const privateFunction: (input: string) => string
        - const publicFunction: (input: number) => string
      - default-const.ts
        - const default: "ðŸŒ³"
      - default-function.ts
        - function <anonymous>(a: number, b: number): string
",
  ],
]
`;

exports[`cli module call with no arguments 1`] = `
[
  [
    "ts-print-tree/
â””â”€â”€ src/
    â”œâ”€â”€ cli.ts
    â”‚   â”œâ”€â”€ const pathFilter: (ignorePatterns: (string | RegExp)[]) => (path: string) => boolean
    â”‚   â”œâ”€â”€ function formatAsTree(node: TreeNode, prefix?: string, isLast?: boolean, isRoot?: boolean): string
    â”‚   â””â”€â”€ function formatAsList(node: TreeNode, depth?: number): string
    â”œâ”€â”€ index.ts
    â”‚   â”œâ”€â”€ interface TreeNode
    â”‚   â””â”€â”€ function tree(rootDir?: string, pathFilter?: (path: string) => boolean, visibilityLevel?: VisibilityLevel): TreeNode
    â””â”€â”€ tests/
        â”œâ”€â”€ class.ts
        â”‚   â”œâ”€â”€ class DefaultClass
        â”‚   â”‚   â”œâ”€â”€ property static foo: string
        â”‚   â”‚   â”œâ”€â”€ property bar: string[]
        â”‚   â”‚   â””â”€â”€ method constructor(baz: number[]): DefaultClass
        â”‚   â”œâ”€â”€ class NamedClass extends DefaultClass
        â”‚   â”‚   â”œâ”€â”€ method get version(): number
        â”‚   â”‚   â”œâ”€â”€ method constructor(): NamedClass
        â”‚   â”‚   â”œâ”€â”€ method double(): void
        â”‚   â”‚   â””â”€â”€ method octuple(): void
        â”‚   â””â”€â”€ const DefaultClass: typeof DefaultClass
        â”œâ”€â”€ consts.ts
        â”‚   â”œâ”€â”€ const publicArray: string[]
        â”‚   â”œâ”€â”€ interface PublicInterface
        â”‚   â””â”€â”€ const publicFunction: (input: number) => string
        â”œâ”€â”€ default-const.ts
        â”‚   â””â”€â”€ const default: "ðŸŒ³"
        â””â”€â”€ default-function.ts
            â””â”€â”€ function <anonymous>(a: number, b: number): string
",
  ],
]
`;
